src/engine.ml

```diff
@@ -101,6 +101,21 @@ let hmac_and_out protocol key { hmac_algorithm; my_hmac; _ } (p : Packet.pkt) =
   let p' = Packet.with_header { header with Packet.hmac } p in
   Packet.encode protocol (key, p')
 
+let encrypt_and_out protocol key { my_key; my_hmac; _ } (p : Packet.pkt) =
+  let to_be_signed = Packet.Tls_crypt.to_be_signed key p in
+  let hmac = Mirage_crypto.Hash.SHA256.hmac ~key:my_hmac to_be_signed in
+  let iv = Cstruct.sub hmac 0 16 in
+  let ctr = Mirage_crypto.Cipher_block.AES.CTR.ctr_of_cstruct iv in
+  let header = Packet.header p in
+  let p = Packet.with_header { header with Packet.hmac } p in
+  let buf, enc_off, enc_len = Packet.Tls_crypt.encode protocol (key, p) in
+  let encrypted =
+    Mirage_crypto.Cipher_block.AES.CTR.encrypt ~key:my_key ~ctr
+      (Cstruct.sub buf enc_off enc_len)
+  in
+  Cstruct.blit encrypted 0 buf enc_off enc_len;
```

Comment:
not obvious to me right now why `encrypted` should fit into the bounds returned from `Packet.Tls_crypt.encode` (`enc_off`, `enc_len`). Maybe worth a comment or a guard?

Reply:
Hmm, do you mean in `(Cstruct.sub buf enc_of enc_len)`? Or `encrypted` having length `enc_len`? The former is by construction in `Packet.Tls_crypt.encode`. The latter is due to AES in CTR-mode not adding any padding or overhead. I can add a comment for both points, but for the former I don't know if I should put it in the function or at the call site(s).

Reply:
I think it is not the easiest code to follow, but I find it hard to make it more clearer. Comments would help!